### PR TITLE
Add support for using JBang

### PR DESCRIPTION
**Pull request by maxandersen**
#Descripton

Adds support for jbang like there are for maven and gradle.

Advantage is you don't need to have extra files and deep structures if you are just doing some pulumi sripting.

Minimal requirement as implemented here is:

```
src/main.java
Pulumi.yaml
jbang.properties
```

Could even cut it down and look for `src/main.java` if no `jbang.properties` found.

Dependencies are done via `//DEPS`in the .java file, i.e.:

```
//DEPS com.pulumi:pulumi:${env.PULUMI_JAVA_SDK_VERSION:0.1.0}
```

or simply:

```
//DEPS com.pulumi:pulumi:0.1.0
```

notes:
Looks for `jbang.properties` as a marker similar to `pom.xml` and `settings.gradle`.
It is not required for jbang to work, thus we could also just look for a standard source file i.e. `src/main.java`

Also handles `binary: myFile.java` as a way to customize which of possible "scripts" to use.

Also allow using Groovy (.groovy) and Kotlin (.kt) via JBang, but that is untested for now.- Fix typo and use GAV that exist in maven central

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by maxandersen** [comment]
Two things I haven't figured out yet:

1) hooking up with the code gen feature. Not sure if relevant/affected

2) how/when "build" command is called by pulumi. Is there a flag or specific sequence of commands to have it call build?

**Comment by AaronFriel** [comment]
Related to this, if we want to add a template: https://github.com/pulumi/templates/pull/283/files

**Comment by maxandersen** [comment]
Yes - I'm happy to complete this just looking for some pointers to my question in the pr description. Thanks!

File: pkg/cmd/pulumi-language-java/main.go [hunk 1]
```diff
@@ -56,15 +57,35 @@ func main() {
...
 	case binary != "":
-		logging.V(3).Infof("language host asked to use specific JAR: `%s`", binary)
-		cmd, err := lookupPath("java")
-		if err != nil {
-			cmdutil.Exit(err)
-		}
-		javaExec, err = newJarExecutor(cmd, binary)
-		if err != nil {
-			cmdutil.Exit(err)
+		logging.V(3).Infof("language host asked to use specific file: `%s`", binary)
```
Comment (t0yv0) [inline pkg/cmd/pulumi-language-java/main.go:60]: "language host asked to use a specific file"

File: pkg/cmd/pulumi-language-java/main.go [hunk 2]
```diff
@@ -56,15 +57,35 @@ func main() {
...
+			cmd, err := probeJBangExecutor()
+			if err != nil {
+				cmdutil.Exit(err)
+			}
+        	javaExec, err = newJBangExecutor(cmd, binary)
+       		if err != nil {
+                cmdutil.Exit(err)
+            }
+		default:
+			cmdutil.Exit(fmt.Errorf("Could not find way to use binary: `%s` with `%s`", binary, suffix))
```
Comment (t0yv0) [inline pkg/cmd/pulumi-language-java/main.go:84]: "Cannot run binary %s with extension %s. Expecting a .jar or a JBang entry-point (.java, .kt, or .groovy)"

File: pkg/cmd/pulumi-language-java/main.go [hunk 3]
```diff
@@ -177,7 +198,7 @@ func (host *javaLanguageHost) determinePulumiPackages(
 	args := host.exec.pluginArgs
 	output, err := host.runJavaCommand(ctx, cmd, args)
 	if err != nil {
-		return nil, errors.Wrapf(err, "language host counld not run plugin discovery command successfully")
+		return nil, errors.Wrapf(err, "language host could not run plugin discovery command successfully")
```
Comment (t0yv0) [inline pkg/cmd/pulumi-language-java/main.go:201]: Nice!

File: pkg/cmd/pulumi-language-java/main.go [hunk 4]
```diff
@@ -35,6 +35,7 @@ func main() {
 	flag.StringVar(&root, "root", "", "Project root path to use")
 	flag.StringVar(&binary, "binary", "", "A relative or an absolute path to a JAR to execute")
```
Comment (t0yv0) [inline pkg/cmd/pulumi-language-java/main.go:36]: "JAR or a JBang entry-point file to execute"

File: pkg/cmd/pulumi-language-java/main.go [hunk 5]
```diff
@@ -35,6 +35,7 @@ func main() {
...
 
+	logging.V(3).Infof("JBANG FTW!")
```
[the remark this answers is not in the document]
Comment (maxandersen) [reply pkg/cmd/pulumi-language-java/main.go:38]: Ha - no we definitely don't :)

File: pkg/cmd/pulumi-language-java/main.go [hunk 6]
```diff
@@ -56,15 +57,35 @@ func main() {
...
+			cmd, err := lookupPath("java")
+			if err != nil {
+				cmdutil.Exit(err)
+			}
+			javaExec, err = newJarExecutor(cmd, binary)
+			if err != nil {
+				cmdutil.Exit(err)
+			}
+		case ".java", ".kt", ".groovy":
+			cmd, err := probeJBangExecutor()
```
Comment (t0yv0) [inline pkg/cmd/pulumi-language-java/main.go:75]: Looks like this pathway is for letting users set `binary:` option in the Pulumi.yaml and have a non-standard entry-point file. Sounds great!

**Comment by t0yv0** [comment]
Thank you for contributing this! I'm just learning about JBang here but I love the idea - it's a great match for short Pulumi programs. Taking it out for a spin shortly.

Please let us know if you can resolve the merge conflicts or need some help with that.

**Comment by t0yv0** [comment]
Also - apologies for CI showing as red.  This is https://github.com/pulumi/pulumi-java/issues/557 - something I will be cleaning up next week.

File: pkg/cmd/pulumi-language-java/main.go (hunk 6 above)
Comment (maxandersen) [reply pkg/cmd/pulumi-language-java/main.go:75]: exactly

**Comment by maxandersen** [comment]
> Please let us know if you can resolve the merge conflicts or need some help with that.

pushed fixes + rebase with main.

**Comment by maxandersen** [comment]
> An executor that does nothing for the "build" command I think should work fine also. It will simply defer the dep fetching until `pulumi up` is called.

so i think it should just work - but didn't find a way to trigger the code thus far.

**Comment by t0yv0** [comment]
Wonderful! Works on my machine, I'll find a way to hitch it up to CI and merge. Thank you!

**Review by t0yv0 — APPROVED** [review]

**Comment by maxandersen** [comment]
Want to merge https://github.com/pulumi/templates/pull/283 too to get matching templates ?

**Comment by t0yv0** [comment]
In principle yes, I'd wait until this change gets released next week though.

**Comment by maxandersen** [comment]
I saw 0.3.0 was released but its release notes does not seem to include this change. Was something missing to make it in?

**Comment by t0yv0** [comment]
Included in https://github.com/pulumi/pulumi-java/blob/main/CHANGELOG.md#030-2022-06-01 ? Where is it not propagating?

**Comment by maxandersen** [comment]
https://github.com/pulumi/pulumi-java/releases latest release don't have it in the change log and it's still in the incoming change log as part of 0.4

**Pull request merged**